File: src/node/generator/lib/network_util.js
const os = require('os');

function getNodeIdFromIp() {
    const interfaces = os.networkInterfaces();
    for (const name of Object.keys(interfaces)) {
        for (const iface of interfaces[name]) {
            if (iface.family === 'IPv4' && !iface.internal) {
                const ipParts = iface.address.split('.');
                if (ipParts.length === 4) {
                    const part3 = parseInt(ipParts[2], 10);
                    const part4 = parseInt(ipParts[3], 10);
                    // Take the last 2 bits of part3 and all 8 bits of part4
                    return ((part3 & 0x03) << 8) | part4;
                }
            }
        }
    }
    return 0;
}

module.exports = { getNodeIdFromIp };

File: src/python/generator/lib/network_util.py
import socket
import psutil

def get_node_id_from_ip():
    try:
        interfaces = psutil.net_if_addrs()
        for interface_name, interface_addresses in interfaces.items():
            if interface_name == 'lo':
                continue
            for address in interface_addresses:
                if address.family == socket.AF_INET:
                    ip = address.address
                    parts = ip.split('.')
                    if len(parts) == 4:
                        part3 = int(parts[2])
                        part4 = int(parts[3])
                        # Take the last 2 bits of part3 and all 8 bits of part4
                        return ((part3 & 0x03) << 8) | part4
    except Exception:
        pass
    return 0
